File: 33 Seach in Rotated Sorted Array/Solution.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        left: int = 0
        right: int = len(nums)-1
        while(left <= right):
            mid: int = (left + right)//2
            if(nums[mid] == target):
                return mid
            else:
                if(nums[mid] < nums[right]):                # right sorted
                    if(nums[mid] < target and target <= nums[right]):
                        left = mid+1
                    else:
                        right = mid-1
                else:                                       #left sorted
                    if(nums[left] <= target and target <nums[mid]):
                        right = mid-1
                    else:
                        left = mid+1
        return -1
    
if __name__ == "__main__":
    # Create an instance of Solution to call the method
    sol = Solution()
    print(sol.search([4, 5, 6, 7, 0, 1, 2], 0))  # Output: 4
    print(sol.search([4, 5, 6, 7, 0, 1, 2], 3))  # Output: -1
    print(sol.search([1], 0))                    # Output: -1

File: 33 Seach in Rotated Sorted Array/Solution.java
class Solution {
    public static int search(int[] nums, int target) {
        int left = 0;
        int right = nums.length - 1;
        
        while(left <= right){
            int mid = (right + left) / 2;
            
            if(nums[mid] == target){
                return mid;
            } else if(nums[mid] < nums[right]){ // Right side is sorted
                if(nums[mid] < target && target <= nums[right]){
                    left = mid + 1;
                } else {
                    right = mid - 1; 
                }
            } else { // Left side is sorted
                if(nums[left] <= target && target < nums[mid]){
                    right = mid - 1;
                } else {
                    left = mid + 1;
                }
            }
        }
        return -1;
    }
    public static void main(String[] args){
        System.out.println(search(new int[]{4,5,6,7,0,1,2}, 0));                    //4
        System.out.println(search(new int[]{4,5,6,7,0,1,2}, 3));                    //-1
        System.out.println(search(new int[]{1}, 0));                                //-1
    }
}